csharp/ql/lib/semmle/code/csharp/security/auth/InsecureDirectObjectReferenceQuery.qll

```diff
@@ -0,0 +1,52 @@
+/** Definitions for the Insecure Direct Object Reference query */
+
+import csharp
+import semmle.code.csharp.dataflow.flowsources.Remote
+import ActionMethods
+
+/**
+ * Holds if `m` is a method that may require checks
+ * against the current user to modify a particular resource.
+ */
+// We exclude admin methods as it may be expected that an admin user should be able to modify any resource.
+// Other queries check that there are authorization checks in place for admin methods.
+private predicate needsChecks(ActionMethod m) { m.isEdit() and not m.isAdmin() }
+
+private Expr getParentExpr(Expr ex) { result = ex.getParent() }
+
+/**
+ * Holds if `m` has a parameter or access a remote flow source
+ * that may indicate that it's used as the ID for some resource
+ */
+private predicate hasIdParameter(ActionMethod m) {
+  exists(RemoteFlowSource src | src.getEnclosingCallable() = m |
+    src.asParameter().getName().toLowerCase().matches(["%id", "%idx"])
+    or
+    // handle cases like `Request.QueryString["Id"]`
+    exists(StringLiteral idStr |
+      idStr.getValue().toLowerCase().matches(["%id", "%idx"]) and
+      getParentExpr*(src.asExpr()) = getParentExpr*(idStr)
```

Comment:
That is probably better, yes. It would have to be local taint tracking, though, as there is otherwise not flow from `Request` to `Request.QueryString`.